src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+            samples (dict): Samples generated by self.get_samples.
+            append_to_shape (list): Specifies additional dimensions for the shape of the results. Defaults to empty list.
+            execution_parameters (`qibolab.ExecutionParameters`): Parameters (nshots,
+                                                        relaxation_time,
+                                                        fast_reset,
+                                                        acquisition_type,
+                                                        averaging_mode)
+
+        Returns:
+            dict: Qibolab results for AcquisitionType.DISCRIMINATION.
+
+        Raises:
+            TypeError: If execution_parameters.acquisition_type is AcquisitionType.RAW or AcquisitionType.INTEGRATION.
+        """
+        shape = [execution_parameters.nshots] + append_to_shape
+        results = {}
+        for ro_pulse in ro_pulse_list:
+            if execution_parameters.acquisition_type is AcquisitionType.DISCRIMINATION:
+                values = np.array(samples[ro_pulse.qubit]).reshape(shape)
+                processed_values = SampleResults(values)
+
+                if execution_parameters.averaging_mode is AveragingMode.SINGLESHOT:
+                    pass
+                elif execution_parameters.averaging_mode is AveragingMode.CYCLIC:
+                    processed_values = (
+                        processed_values.average
+                    )  # generaetes AveragedSampleResults
+
+            elif execution_parameters.acquisition_type is AcquisitionType.RAW:
+                raise TypeError("Emulator does not support raw measurement")
+            elif execution_parameters.acquisition_type is AcquisitionType.INTEGRATION:
+                raise TypeError("Emulator does not support integrated measurement")
+
+            results[ro_pulse.qubit] = results[ro_pulse.serial] = processed_values
+        return results
+
+    def run_pulse_simulation(
+        self,
+        sequence: PulseSequence,
+        instant_measurement: bool = True,
+    ) -> tuple[np.ndarray, list]:
+        """Simulates the input pulse sequence.
+
+        Args:
+            sequence (`qibolab.pulses.PulseSequence`): Pulse sequece to simulate.
+            instant_measurement (bool): Collapses readout pulses to duration of 1 if True. Defaults to True.
+
+        Returns:
+            tuple: Reduced density matrix of final state specified by readout qubits as well as the list of readout qubits in, both in little endian order.
+        """
+        # reduces measurement time to 1 dt to save simulation time
+        if instant_measurement:
+            for i in range(len(sequence)):
+                if type(sequence[i]) is ReadoutPulse:
+                    sequence[i].duration = 1
+
+        # extract waveforms from pulse sequence
+        channel_waveforms = ps_to_waveform_dict(
+            sequence,
+            self.platform2simulator_channels,
```

Comment:
I am not sure if there is a special rule about this, but I would change to a more explicit name, for example
```suggestion
            self.platform_to_simulator_channels,
```
if that complicated attribute is really needed. Otherwise just a simpler name.

Reply:
I don't have a better idea so we'll keep to this for now.